syntax = "proto3";
package neurogenesis;

service Env {
    rpc Reset(ResetRequest) returns (ResetResponse);
    rpc Step(StepRequest) returns (StepResponse);
    rpc Seed(SeedRequest) returns (SeedResponse);
    rpc Spec(SpecRequest) returns (SpecResponse);   
}

message Observation { 
    bytes tensor = 1; 
    repeated int32 shape = 2; 
    string dtype = 3;
}

message Action {
    bytes tensor = 1;
    repeated int32 shape = 2;
    string dtype = 3;
}

message ResetRequest {}

message ResetResponse {
    Observation observation = 1;
}

message StepRequest {
    Action action = 1;
}

message StepResponse {
    Observation  observation = 1;
    double reward = 2;
    bool done = 3;
}

message SeedRequest {
    int64 seed = 1;
}

message SeedResponse {

}

message SpaceSpec {
    repeated int32 shape = 1;
    string dtype = 2;
}

message SpecRequest {}

message SpecResponse {
    SpaceSpec observation = 1;
    SpaceSpec action = 2;
}